File: src/main/java/io/github/saneea/feature/time/Converter.java
package io.github.saneea.feature.time;

import java.time.ZonedDateTime;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import io.github.saneea.Feature;
import io.github.saneea.Feature.Util.IOConsumer;
import io.github.saneea.FeatureContext;
import io.github.saneea.feature.time.format.Format;
import io.github.saneea.feature.time.format.FormatFactory;

public class Converter implements//
		Feature, //
		Feature.CLI, //
		Feature.In.Text.String, //
		Feature.Out.Text.String {

	private static final String IN_FORMAT = "inputFormat";
	private static final String OUT_FORMAT = "outputFormat";

	private String in;
	private IOConsumer<String> out;
	private CommandLine commandLine;

	private final FormatFactory formatFactory = new FormatFactory();

	@Override
	public String getShortDescription() {
		return "convert time from original format to another one";
	}

	@Override
	public void run(FeatureContext context) throws Exception {
		Format inFormat = getFormatFromCLI(IN_FORMAT);
		Format outFormat = getFormatFromCLI(OUT_FORMAT);

		ZonedDateTime timeAsZoned = inFormat.parse(in);
		String converted = outFormat.render(timeAsZoned);

		out.accept(converted);
	}

	private Format getFormatFromCLI(String cliOptionName) {
		return formatFactory.createFormat(commandLine.getOptionValue(cliOptionName));
	}

	@Override
	public void setIn(String in) {
		this.in = in;
	}

	@Override
	public void setOut(IOConsumer<String> out) {
		this.out = out;
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

	@Override
	public Option[] getOptions() {
		Option[] options = { //
				createFormatCliOption("if", IN_FORMAT, "input time format"), //
				createFormatCliOption("of", OUT_FORMAT, "output time format")//
		};
		return options;
	}

	private Option createFormatCliOption(String shortName, String longName, String description) {
		return Option//
				.builder(shortName)//
				.longOpt(longName)//
				.hasArg(true)//
				.argName(formatFactory.getAvailableFormatsString())//
				.required(true)//
				.desc(description)//
				.build();
	}

}

File: src/main/java/io/github/saneea/dvh/feature/time/Converter.kt
package io.github.saneea.dvh.feature.time

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.StringConsumer
import io.github.saneea.dvh.feature.time.format.FormatFactory
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Option

private const val IN_FORMAT = "inputFormat"
private const val OUT_FORMAT = "outputFormat"

class Converter :
    Feature,
    Feature.CLI,
    Feature.CLI.Options,
    Feature.In.Text.String,
    Feature.Out.Text.String {

    override lateinit var inTextString: String
    override lateinit var outTextString: StringConsumer
    override lateinit var commandLine: CommandLine
    private val formatFactory = FormatFactory()

    override val meta = Meta("convert time from original format to another one")

    override fun run() {
        val inFormat = getFormatFromCLI(IN_FORMAT)
        val outFormat = getFormatFromCLI(OUT_FORMAT)
        val timeAsZoned = inFormat.parse(inTextString)
        val converted = outFormat.render(timeAsZoned)
        outTextString(converted)
    }

    private fun getFormatFromCLI(cliOptionName: String) =
        formatFactory.createFormat(commandLine.getOptionValue(cliOptionName))

    override val options
        get() = arrayOf(
            createFormatCliOption("if", IN_FORMAT, "input time format"),
            createFormatCliOption("of", OUT_FORMAT, "output time format")
        )

    private fun createFormatCliOption(shortName: String, longName: String, description: String) =
        Option
            .builder(shortName)
            .longOpt(longName)
            .hasArg(true)
            .argName(formatFactory.availableFormatsString)
            .required(true)
            .desc(description)
            .build()
}